samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,889 @@
+
+	memcpy(data_frame + 9, payload, length);
+
+	int frame_size = 9 + length;
+
+	if (sendall(socket_fd, (const char *)data_frame, frame_size) < 0) {
+		LOG_ERR("ERROR writing to socket");
+		return;
+	}
+}
+
+const char *get_frame_type_name(unsigned char type)
+{
+	switch (type) {
+	case 0:
+		return "DATA";
+	case 1:
+		return "HEADERS";
+	case 2:
+		return "PRIORITY";
+	case 3:
+		return "RST_STREAM";
+	case 4:
+		return "SETTINGS";
+	case 5:
+		return "PUSH_PROMISE";
+	case 6:
+		return "PING";
+	case 7:
+		return "GOAWAY";
+	case 8:
+		return "WINDOW_UPDATE";
+	case 9:
+		return "CONTINUATION";
+	default:
+		return "UNKNOWN";
+	}
+}
+
+void print_http2_frames(struct http2_frame *frames, unsigned int frame_count)
+{
+	const char *bold = "\033[1m";
+	const char *reset = "\033[0m";
+	const char *green = "\033[32m";
+	const char *blue = "\033[34m";
+
+	for (unsigned int i = 0; i < frame_count; i++) {
+		printf("%s=====================================%s\n", green,
+			reset);
+		printf("%sReceived %s Frame :%s\n", bold,
+			get_frame_type_name(frames[i].type), reset);
+		printf("  %sLength:%s %u\n", blue, reset, frames[i].length);
+		printf("  %sType:%s %u (%s)\n", blue, reset, frames[i].type,
+			get_frame_type_name(frames[i].type));
+		printf("  %sFlags:%s %u\n", blue, reset, frames[i].flags);
+		printf("  %sStream Identifier:%s %u\n", blue, reset,
+			frames[i].stream_identifier);
+		printf("  %sPayload:%s ", blue, reset);
+		for (unsigned int j = 0; j < frames[i].length; j++)
+			printf("%02x ", frames[i].payload[j]);
+
+		printf("\n");
+		printf("%s=====================================%s\n\n\n", green,
+			reset);
+	}
+}
+
+unsigned int parse_http2_frames(unsigned char *buffer, unsigned long buffer_len,
+	struct http2_frame *frames)
+{
+	unsigned int frame_count = 0;
+	unsigned long pos = 0;
+
+	while (pos + HTTP2_FRAME_HEADER_SIZE <= buffer_len &&
+		frame_count < MAX_FRAMES) {
+		frames[frame_count].length =
+			(buffer[pos + HTTP2_FRAME_LENGTH_OFFSET] << 16) |
+			(buffer[pos + HTTP2_FRAME_LENGTH_OFFSET + 1] << 8) |
+			buffer[pos + HTTP2_FRAME_LENGTH_OFFSET + 2];
+		frames[frame_count].type =
+			buffer[pos + HTTP2_FRAME_TYPE_OFFSET];
+		frames[frame_count].flags =
+			buffer[pos + HTTP2_FRAME_FLAGS_OFFSET];
+		frames[frame_count].stream_identifier =
+			(buffer[pos + HTTP2_FRAME_STREAM_ID_OFFSET] << 24) |
+			(buffer[pos + HTTP2_FRAME_STREAM_ID_OFFSET + 1] << 16) |
+			(buffer[pos + HTTP2_FRAME_STREAM_ID_OFFSET + 2] << 8) |
+			buffer[pos + HTTP2_FRAME_STREAM_ID_OFFSET + 3];
+		frames[frame_count].stream_identifier &= 0x7FFFFFFF;
+
+		pos += HTTP2_FRAME_HEADER_SIZE;
+
+		/* Check if payload size doesn't exceed buffer size */
+		if (pos + frames[frame_count].length > buffer_len)
+			break;
+
+		frames[frame_count].payload = buffer + pos;
+
+		pos += frames[frame_count].length;
+
+		frame_count++;
+	}
+
+	return frame_count;
+}
+
+bool is_first_flag_set(unsigned char flags) { return (flags & 0x01) != 0; }
```

Comment:
Also please note, that generally it'd be recommended to create defines/enums for all constants specified in the spec (flags, frame types, setting types) and use them in the code, instead of using magic numbers in various places.